raft panic: compact 394384 out of bounds [454490:394384]

It happens on AWS 3 m3.large machine cluster.

```
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 rafthttp: stream client with e7333b0ad35b3e9 in term 4 is over
2014/11/24 06:08:03 rafthttp: start stream client with e7333b0ad35b3e9 in term 4
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
2014/11/24 06:08:03 sender: reach the maximal serving to http://172.31.39.94:7001/raft
panic: compact 394384 out of bounds [454490:394384]

goroutine 25 [running]:
runtime.panic(0x6e8760, 0xc208817e40)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/raft.(*raftLog).compact(0xc20804a5a0, 0x60490, 0x0)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/log.go:201 +0x145
github.com/coreos/etcd/raft.(*raft).compact(0xc20803c840, 0x60490, 0xc2080f8880, 0x3, 0x3, 0xc208e52000, 0x2cec3, 0x495d5)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:536 +0x254
github.com/coreos/etcd/raft.(*node).run(0xc208018c80, 0xc20803c840)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:275 +0x6be
created by github.com/coreos/etcd/raft.StartNode
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:162 +0x4f8
```

--- a/raft/raft.go
+++ b/raft/raft.go
@@ -534,6 +534,10 @@ func (r *raft) compact(index uint64, nodes []uint64, d []byte) {
 	if index > r.raftLog.applied {
 		panic(fmt.Sprintf("raft: compact index (%d) exceeds applied index (%d)", index, r.raftLog.applied))
 	}
+	if index < r.raftLog.offset {
+		//TODO: return an error?
+		return
+	}
 	r.raftLog.snap(d, index, r.raftLog.term(index), nodes)
 	r.raftLog.compact(index)
 }

